pkg/provider/filters.go

```diff
@@ -115,3 +123,78 @@ func (env *TestEnvironment) GetHugepagesPods() []*Pod {
 	}
 	return filteredPods
 }
+
+func (env *TestEnvironment) GetCPUPinningPodsWithDpdk() []*Pod {
+	return filterCPUPinningPods(filterDPDKRunningPods(env))
+}
+
+func filterDPDKRunningPods(env *TestEnvironment) []*Pod {
+	var filteredPods []*Pod
+	const (
+		dpdkDriver           = "vfio-pci"
+		findDeviceSubCommand = "find . -name"
+	)
+	o := clientsholder.GetClientsHolder()
+	for _, pod := range env.Pods {
+		if len(pod.MultusPCIs) == 0 {
+			continue
+		} else {
+			ctx := clientsholder.NewContext(pod.Namespace, pod.Name, pod.Spec.Containers[0].Name)
+			findCommand := fmt.Sprintf("%s '%s'", findDeviceSubCommand, pod.MultusPCIs[0])
+			outStr, errStr, err := o.ExecCommandContainer(ctx, findCommand)
+			if err != nil || errStr != "" {
+				logrus.Errorf("Failed to execute command %s in debug %s, errStr: %s, err: %s", findCommand, pod.String(), errStr, err)
+				continue
+			}
+			if strings.Contains(outStr, dpdkDriver) {
+				filteredPods = append(filteredPods, pod)
+			}
+		}
+	}
+	return filteredPods
+}
+
+func filterCPUPinningPods(pods []*Pod) []*Pod {
+	const (
+		verifyCPUAffinityCommand = "for pid in $(ps -A -o pid=); do taskset -pc $pid 2>/dev/null; done"
+		newLineCharacter         = "\n"
+		spaceCharacter           = " "
+	)
+
+	var filteredPods []*Pod
+
+	if len(pods) == 0 {
+		return filteredPods
+	}
+
+	o := clientsholder.GetClientsHolder()
+
+	for _, pod := range pods {
+		ctx := clientsholder.NewContext(pod.Namespace, pod.Name, pod.Spec.Containers[0].Name)
+		outStr, errStr, err := o.ExecCommandContainer(ctx, verifyCPUAffinityCommand)
+		if outStr == "" {
```

Comment:
Why not just `if outStr == "" && err != nil && errStr != "" {` ?

Other than that, although it might not be the case here, just note that stdout and stderr are not mutually exclusive. Some programs will print you output in both streams. I recommend this pattern here:
```
outStr, errStr, err := o.ExecCommandContainer(ctx, verifyCPUAffinityCommand)
if err != nil || errStr != "" {
  logrus.Errorf("failed to execute command...", pod, command, errStr, outStr, err) // see I also added outStr here as extra debugging info.
  continue
}

lines := strings.Split(outStr, newLineCharacter)
```

Reply:
Thanks for the recommendation, but it is not applicable here. The command executed in the container would give error each time and that's why the error output is sent to null device. So, the output of the command has to be checked at first.

Reply:
Why not `if outStr == "" && err != nil && errStr != "" {` -> I did not choose this for the better readability of the flow. The focus is more into the outputStr.

Reply:
Ok, I just tried to convey that, in go, after invoking any function that can return any error (along with other values), it's idiomatic to check the error variable first. As a general rule, if err is not nil, the rest of the returned values are usually expected to be zeroed/empty/nil. See go's [builtin os.ReadDir](https://github.com/golang/go/blob/939f9fd64a740ef770a0ca3588f8c193dceebe83/src/os/dir.go#L118 ) function.

The `ExecCommandContainer()` function could also fail because of [errors occurred even before](https://github.com/test-network-function/cnf-certification-test/blob/a55d59cf11149b3895b3ff0a8996ea704d4fb707/internal/clientsholder/command.go#L60) the command can actually run inside the container: networking issues, the container has just stopped or is in a CrashLoopBack phase... In those cases, I don't see quite useful to check stdOut first.

Reply:
Thanks, it seems to me that you did not run the command that we are executing it. It's simply not applicable to check err at first, I know about the best practices but we have to check result at first to track the valid error. I would suggest to run the `verifyCPUAffinityCommand` inside a container to test it. Btw, I do not think that we would need this piece of code anymore as we would probably not check affinity with `taskset`.